File: frontend/src/services/contactService.js
import apiClient from './api';

export const contactService = {
  /**
   * Fetch all emergency contacts for current user.
   * @param {string} search - Optional search query
   * @returns {Promise<Array>} List of contact objects
   */
  async getContacts(search = '') {
    const params = search ? { search } : {};
    const response = await apiClient.get('/contacts/', { params });
    return response.data;
  },

  /**
   * Create a new emergency contact.
   * @param {Object} contactData - { contact_name, phone_number, relationship }
   * @returns {Promise<Object>} Created contact object
   */
  async createContact(contactData) {
    const response = await apiClient.post('/contacts/', contactData);
    return response.data;
  },

  /**
   * Update an existing emergency contact.
   * @param {number} id - Contact ID
   * @param {Object} contactData - { contact_name, phone_number, relationship }
   * @returns {Promise<Object>} Updated contact object
   */
  async updateContact(id, contactData) {
    const response = await apiClient.put(`/contacts/${id}`, contactData);
    return response.data;
  },

  /**
   * Delete an emergency contact by ID.
   * @param {number} id - Contact ID
   * @returns {Promise<void>}
   */
  async deleteContact(id) {
    await apiClient.delete(`/contacts/${id}`);
  },
};

export default contactService;

File: backend/app/services/contact_service.py
from typing import List, Optional
from fastapi import HTTPException, status
from sqlalchemy.orm import Session

from app.repositories.contact_repository import ContactRepository
from app.schemas.contact import ContactCreate, ContactUpdate, ContactResponse
from app.models.contact import EmergencyContact


class ContactService:
    """Service handling business rules and authorization for emergency contacts."""

    def __init__(self, db: Session):
        self.contact_repo = ContactRepository(db)

    def list_contacts(self, user_id: int, search: Optional[str] = None) -> List[ContactResponse]:
        """List all emergency contacts belonging to user with optional search."""
        contacts = self.contact_repo.get_by_user_id(user_id, search)
        return [ContactResponse.model_validate(c) for c in contacts]

    def get_contact(self, contact_id: int, user_id: int) -> ContactResponse:
        """Get contact details by ID with ownership check."""
        contact = self.contact_repo.get_by_id(contact_id, user_id)
        if not contact:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Emergency contact not found or access unauthorized."
            )
        return ContactResponse.model_validate(contact)

    def create_contact(self, contact_in: ContactCreate, user_id: int) -> ContactResponse:
        """Create new emergency contact for user."""
        contact = self.contact_repo.create(contact_in, user_id)
        return ContactResponse.model_validate(contact)

    def update_contact(self, contact_id: int, contact_in: ContactUpdate, user_id: int) -> ContactResponse:
        """Update existing emergency contact with ownership verification."""
        contact = self.contact_repo.get_by_id(contact_id, user_id)
        if not contact:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Emergency contact not found or access unauthorized."
            )
        updated_contact = self.contact_repo.update(contact, contact_in)
        return ContactResponse.model_validate(updated_contact)

    def delete_contact(self, contact_id: int, user_id: int) -> None:
        """Delete emergency contact with ownership verification."""
        contact = self.contact_repo.get_by_id(contact_id, user_id)
        if not contact:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Emergency contact not found or access unauthorized."
            )
        self.contact_repo.delete(contact)
